### PR TITLE
feat: add GetOrCreate and GetOrCall methods.

### PR DESCRIPTION
**Pull request by DmitriyMV**
Also add Len/Clear/Trunc utility methods.

Signed-off-by: Dmitriy Matrenichev <dmitry.matrenichev@siderolabs.com>

### CODE REVIEW

File: containers/map.go [hunk 1]
```diff
@@ -22,6 +22,49 @@ func (m *ConcurrentMap[K, V]) Get(key K) (V, bool) {
 	return val, ok
 }
 
+// GetOrCreate returns the existing value for the key if present. Otherwise, it stores and returns the given value.
+// The loaded result is true if the value was loaded, false if stored.
+func (m *ConcurrentMap[K, V]) GetOrCreate(key K, val V) (V, bool) {
+	m.mx.Lock()
```
Comment (utkuozdemir) [inline containers/map.go:28]: double-checked locking here?

Comment (DmitriyMV) [reply containers/map.go:28]: Why? We only access map if the lock was held.

Comment (utkuozdemir) [reply containers/map.go:28]: why not? if there is already an element, we don't even need to acquire the lock, no? or doesn't it work that way in Go for some reason?

File: containers/map.go [hunk 2]
```diff
@@ -22,6 +22,49 @@ func (m *ConcurrentMap[K, V]) Get(key K) (V, bool) {
...
+
+	return val, false
+}
+
+// GetOrCall returns the existing value for the key if present. Otherwise, it calls fn, stores the result and returns it.
+// The loaded result is true if the value was loaded, false if it was created using fn.
+//
+// The main reason for this function is to avoid unnecessary allocations if you use pointer types as values, since
+// compiler cannot prove that the value does not escape if it's not stored.
+func (m *ConcurrentMap[K, V]) GetOrCall(key K, fn func() V) (V, bool) {
```
Comment (smira) [inline containers/map.go:49]: can we implement stockpiling protection here? or you're looking for some other layer?

File: containers/map.go (hunk 1 above)
Comment (DmitriyMV) [reply containers/map.go:28]: No, if we try to read from the Go `map` that is currently being written, this is going to break with a huge stacktrace.

Comment (utkuozdemir) [reply containers/map.go:28]: hmm, interesting, but ok :)

File: containers/map.go (hunk 2 above)
Comment (DmitriyMV) [reply containers/map.go:49]: `fn` should usually be just a variable creation, not a _heavy_ operation. If I understand you correctly - I'm going to demonstrate how to use it omni.

**Review by utkuozdemir — APPROVED** [review]

File: containers/map.go (hunk 2 above)
Comment (smira) [reply containers/map.go:49]: In Omni e.g. `kubeconfig` generation is a _heavy_ operation, so probably it should be implemented with stockpile protection.

File: containers/map.go [hunk 3]
```diff
@@ -65,3 +131,11 @@ func (m *ConcurrentMap[K, V]) Clear() {
 		delete(m.m, k)
 	}
 }
+
+// Trunc resets the underlying map.
+func (m *ConcurrentMap[K, V]) Trunc() {
```
Comment (smira) [inline containers/map.go]: `Reset` might be a better name

**Review by smira — APPROVED** [review]

**Pull request merged**